### PR TITLE
Add nix flake

### PR DESCRIPTION
**Pull request by billyb2**
This makes setting up the dependencies for building the project much easier. To use, you need to run:
`nix build '.?submodules=1'`

### CODE REVIEW

**Comment by billyb2** [comment]
If this PR is accepted, is there a docs repo I can make a follow up to?

File: flake.nix [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  description = " A fast usermode x86 and x86-64 emulator for Arm64 Linux";
+
+  inputs = {
+    nixpkgs.url = "github:nixos/nixpkgs?ref=nixos-unstable";
+    flake-utils.url = "github:numtide/flake-utils";
+  };
+
+  outputs = { self, nixpkgs, flake-utils, ... }:
+    let
+      supportedSystems = [ "x86_64-linux" "aarch64-linux" ];
```
Comment (alyssarosenzweig) [inline flake.nix:11]: why x86?

File: flake.nix [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+          fex = pkgs.stdenv.mkDerivation rec {
+            name = "FEXBash";
+            src = self;
+            nativeBuildInputs = deps ++ [ pkgs.stdenv.cc.cc.lib ];
+            dontUseCmakeConfigure = true;
+            dontUseCmakeInstall = true;
+            configurePhase = ''
+              mkdir Build
+              cd Build
+              CC=clang CXX=clang++ cmake -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_BUILD_TYPE=Release -DUSE_LINKER=lld -DENABLE_LTO=True -DBUILD_TESTS=False -DENABLE_ASSERTIONS=False -G Ninja ..
```
Comment (alyssarosenzweig) [inline flake.nix:45]: Is the prefix right for nix...?

File: flake.nix [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            '';
+            buildPhase = ''
+              ninja -j $(nproc)
```
Comment (alyssarosenzweig) [inline flake.nix:48]: I think ninja should default to nproc so just `ninja` probably works

File: flake.nix [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        let
+          pkgs = nixpkgsFor.${system};
+          deps = with pkgs; [
+            git
+            cmake
+            ninja
+            clang
+            binutils
+            lld
+            llvm
```
Comment (alyssarosenzweig) [inline flake.nix:26]: do we depend on llvm proper or just clang and lld?

File: flake.nix [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            gcc-unwrapped
+            python311Packages.setuptools
+          ];
+        in {
+          fex = pkgs.stdenv.mkDerivation rec {
+            name = "FEXBash";
+            src = self;
+            nativeBuildInputs = deps ++ [ pkgs.stdenv.cc.cc.lib ];
+            dontUseCmakeConfigure = true;
+            dontUseCmakeInstall = true;
```
Comment (Sonicadvance1) [inline flake.nix:41]: Why not use cmake's install step?

File: flake.nix [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+              mkdir Build
+              cd Build
+              CC=clang CXX=clang++ cmake -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_BUILD_TYPE=Release -DUSE_LINKER=lld -DENABLE_LTO=True -DBUILD_TESTS=False -DENABLE_ASSERTIONS=False -G Ninja ..
+            '';
+            buildPhase = ''
+              ninja -j $(nproc)
+            '';
+            installPhase = ''
+              mkdir -p $out/bin
+              mv ./Bin/* $out/bin/
```
Comment (Sonicadvance1) [inline flake.nix:52]: We have more install components than just what is in Bin/. Also some of the binaries in Bin shouldn't be installed as some are utility applications used during the build process.
Missing files would include config files, man pages, thunk libraries, and the like.
Would using the cmake install step resolve this?

File: flake.nix [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+            libepoxy
+            nasm
+            pkg-config
+            python3
+            gcc-unwrapped
+            python311Packages.setuptools
+          ];
+        in {
+          fex = pkgs.stdenv.mkDerivation rec {
+            name = "FEXBash";
```
Comment (Sonicadvance1) [inline flake.nix:37]: Is this the package name? FEXBash is just one utility application. If anything it should match our PPA package name of `fex-emu` https://github.com/FEX-Emu/FEX-ppa/blob/main/deb_base/control#L57

File: flake.lock
Comment (Sonicadvance1) [inline flake.lock:1]: As someone that has never used nix. What's a flake lock file for?

**Review by Sonicadvance1 — COMMENTED** [review]
What an interesting package description format.

**Comment by Sonicadvance1** [comment]
> If this PR is accepted, is there a docs repo I can make a follow up to?

We also have a `docs/` directory in this repo if a markdown file is enough there

File: flake.nix [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+            SDL2
+            libepoxy
+            nasm
+            pkg-config
+            python3
+            gcc-unwrapped
+            python311Packages.setuptools
+          ];
+        in {
+          fex = pkgs.stdenv.mkDerivation rec {
```
Comment (bylaws) [inline flake.nix:36]: Can't you use clangStdenv here to avoid the need to set cc and manually install llvm?

**Comment by neobrain** [comment]
Closing due to inactivity.

We can reopen it if interest picks up again, though personally I don't currently see a net benefit in adding poorly maintained build system integrations into `main`.

**Pull request closed without merging**